### PR TITLE
feature: paginated response handler

### PR DESCRIPTION
**Pull request by aaronenberg**
`get_collections` does the paginated response handler logic and `Page` models a paginated response. Using generics we can accommodate any type of collection e.g. comments, tracks, playlists.

### CODE REVIEW

**Review by maxjoehnk — CHANGES_REQUESTED** [review]
I don't think this is a good idea for paging as the user has no control over the amount of data returned. This may lead to rate limit issues.
I'm currently thinking about a paging api for another application and think that maybe returning a `futures::Stream` would be a good idea.
We could set a page size in the request builder and then return a stream of pages (e.g. `Stream<Item = Vec<T>>`)
This allows the user to e.g. get the first 2 pages (`client.tracks().stream().take(2).collect().await?`) or skip the first page because it already got requested.
I'm not yet sure how feasible this kind of api is, but I think we should give the user as much flexibility as possible.
I would also like to keep the current get methods as is and add the paging api on top (e.g. with a `.stream()` method)

File: src/user.rs
Comment (maxjoehnk) [inline src/user.rs]: I've changed the api to only return a `Vec` as the `Option` doesn't provide any benefit to the user.

**Comment by aaronenberg** [comment]
Found [this](http://xion.io/post/code/rust-unfold-pagination.html#fnref:1) example of using `futures::stream::unfold` for pagination. I'll see if I can get something like this working. Let me know if you like this idea.

> This allows the user to e.g. get the first 2 pages (client.tracks().stream().take(2).collect().await?) or skip the first page because it already got requested.

Yeah that would be cool if we can cache `next_href` between `.stream()` calls.

**Comment by maxjoehnk** [comment]
> Yeah that would be cool if we can cache next_href between .stream() calls.

I think we should skip caching of `next_href` for now and add it later as it should not impact the public api but introduces additional complexity to this pr.

**Comment by aaronenberg** [comment]
Hey, I finally had some time in my schedule to work on this.
The new logic allows the client to return a `Stream` of any soundcloud entity. I basically copy/pasted it from [these guys](https://github.com/softprops/hubcaps).

You've might have seen recently, Soundcloud wrote a blog [post](https://developers.soundcloud.com/blog/pagination-updates-on-our-api) announcing they are updating more endpoints to use the new pagination API. When that is complete, the `/users/{id}/tracks` endpoint, which I've included here, will paginate correctly. Before adding all the other endpoints to the pagination I wanted to get your approval on this approach.

File: src/client.rs
Comment (maxjoehnk) [inline src/client.rs:27]: Why have you added the host property here? You don't seem to be using it

File: src/client.rs
Comment (maxjoehnk) [inline src/client.rs]: We should never unwrap `Result`s in library code.
Is there a reason for the changed method signature (`reqwest::Result` instead of `Result`)?

File: src/client.rs
Comment (maxjoehnk) [inline src/client.rs]: Same as above.

File: src/client.rs
Comment (maxjoehnk) [inline src/client.rs]: You should be able to use the `json` method here instead of deserializing manually.

File: src/client.rs
Comment (maxjoehnk) [inline src/client.rs]: Can you re-add the 3 tests you've removed?

File: src/user.rs
Comment (maxjoehnk) [inline src/user.rs]: Are you going to implement the playlists method with the new stream api?

**Review by maxjoehnk — CHANGES_REQUESTED** [review]
I like how the api turned out. I would prefer if you could also add a way to just get the first page (so instead of calling `iter().take(15)` just call `.get()`).

Also I don't really like the re-export of the Future and Stream Types, maybe you can just use `BoxStream` from the futures crate?

File: src/user.rs
Comment (aaronenberg) [reply src/user.rs]: yes, I wanted your approval before implementing

File: src/client.rs
Comment (aaronenberg) [reply src/client.rs]: I replaced `test_get_collections` and `test_get_user_tracks` with `test_paginate_user_tracks` and added `test_get_first_page_user_tracks`. I also removed `test_get_user_playlists` since it's the same logic with the new API. I'll add it back if having the coverage is important to you. I also removed `test_get_user_tracks_from_permalink` because that code is already covered with `test_get_user_permalink` and `test_paginate_user_tracks`.

File: src/client.rs
Comment (aaronenberg) [reply src/client.rs]: Was trying to minimize the sum total of changes in this pull request. Adding it back.

File: src/client.rs
Comment (aaronenberg) [reply src/client.rs:27]: Fixed that. I think it's nicer delegating to the client the work of prepending the host name to URIs.

File: src/client.rs
Comment (aaronenberg) [reply src/client.rs]: Nice

**Comment by aaronenberg** [comment]
> I like how the api turned out. I would prefer if you could also add a way to just get the first page (so instead of calling `iter().take(15)` just call `.get()`).

Can we combine these ideas? I think It'd be cool to be able to get the first n pages with `.get(..., n)`.
 
> Also I don't really like the re-export of the Future and Stream Types, maybe you can just use `BoxStream` from the futures crate?

Thanks. Didn't even know those types existed.

**Comment by aaronenberg** [comment]
I've added just about every GET resource endpoint (except /me) to the stream api. If there's anything I missed let me know.

File: src/comment.rs
Comment (maxjoehnk) [inline src/comment.rs]: You are not using the extra property currently so this could be removed right?

File: src/comment.rs
Comment (maxjoehnk) [inline src/comment.rs]: I don't really like how the parts of the url are joined here as it won't work when you add another query parameter. Maybe you could use `Url::parse_with_params` instead?

File: src/track.rs
Comment (maxjoehnk) [inline src/track.rs]: I think we should either use different naming for the comment types or represent it as an enum.
```rust
enum Comments {
  User { client, user_id },
  Track { client, track_id }
}
```

You could share the `iter` and `get` implementation and just have specific `path` implementations then.

**Review by maxjoehnk — CHANGES_REQUESTED** [review]
I think we should introduce a trait for this recurring pattern of a `path` method specifing where to fetch data and an `iter` and `get` method for the data fetching.
The trait could have a base implementation for both methods so all these objects only need to implement the path method.

If you don't mind I can do these changes for you.

File: src/client.rs
Comment (maxjoehnk) [inline src/client.rs]: You don't need to clone the token here. The following should work, if not you can try `self.auth_token.as_ref()` instead of `Some(ref token)`.

```suggestion
        if let Some(ref token) = self.auth_token {
            headers.insert(
                reqwest::header::AUTHORIZATION,
                format!("OAuth {}", token).parse().unwrap(),
            );
        }
```

File: src/client.rs
Comment (maxjoehnk) [reply src/client.rs]: I would actually prefer having `test_get_user_playlists` as the logic might be the same but there can still be issues with deserializing or the url we're accessing.

**Comment by maxjoehnk** [comment]
I really like how this API has evolved. This should be ready to merge when we've reduced the redundancy.
Thank you for this contribution.

File: src/track.rs
Comment (aaronenberg) [reply src/track.rs]: Yes, I like this much better.

**Comment by aaronenberg** [comment]
Awesome! Thanks for letting me contribute. This has been an invaluable learning experience for me. If you'd like me to take on any of the changes you've mentioned I'll be happy to.

File: src/comment.rs
Comment (maxjoehnk) [reply src/comment.rs]: Apparently `Url::parse_with_params` doesn't like url's without a host so I had to use string concatenation as before but without the `Vec` as it's the only extension for the base url.

**Comment by maxjoehnk** [comment]
I did the changes and added some tests and everything seems to be working fine.
I'm still thinking about some api changes but we can merge this for now and open a new PR for further changes.

I would like to reorganize the code and the module structure as some files have grown beyond my preferences.
I also think we should rename the `iter` method to `stream` and let `get` return a future instead of a stream.
Not sure whether `num_pages` should be changed to `count` then.

Do you have anything you want to change regarding this PR? If not I'll go ahead and merge this.

**Comment by aaronenberg** [comment]
Changes look great. I have nothing more to add to this PR. Cheers!

**Pull request merged**
